perllib/FixMyStreet/App/Controller/Around.pm

```diff
@@ -221,6 +238,52 @@ sub check_location_is_acceptable : Private {
     return $c->forward('/council/load_and_check_areas');
 }
 
+=head2 check_and_stash_category
+
+Check that the 'category' query param is valid, if it's present. Stores
+the validated string in the stash as filter_category.
+Puts all the valid categories in filter_categories on the stash.
+
+=cut
+
+sub check_and_stash_category : Private {
+    my ( $self, $c ) = @_;
+
+    my $all_areas = $c->stash->{all_areas};
+    my @bodies = $c->model('DB::Body')->search(
+        { 'body_areas.area_id' => [ keys %$all_areas ], deleted => 0 },
+        { join => 'body_areas' }
+    )->all;
+    my %bodies = map { $_->id => $_ } @bodies;
+
+    my @contacts = $c->model('DB::Contact')->not_deleted->search(
+        {
+            body_id => [ keys %bodies ],
+        },
+        {
+            columns => [ 'category' ],
+            order_by => [ 'category' ],
+            distinct => 1
+        }
+    )->all;
+    my @categories = map { $_->category } @contacts;
+    $c->stash->{filter_categories} = \@categories;
+
+
+    my $category = $c->req->param('category');
+    if ( $category ) {
+        my $count = $c->model('DB::Contact')->not_deleted->search(
+            {
+                body_id => [ keys %bodies ],
+                category => $category
+            }
+        )->count;
```

Comment:
As you've just looked up all the categories above, could this be done looking at `@categories` rather than another DB call?